prov/efa/docs/efa_rdm_protocol_v4.md

```diff
@@ -0,0 +1,1307 @@
+REQ optional header and application data. The binary format of EAGER_RTW mandatory header is listed
+in table 3.8:
+
+Table: 3.8 the binary format of EAGER_RTW packet's mandatory header
+
+| Name | Length (bytes) | type | C language type | Note |
+|-|-|-|-|-|
+| `type`           | 1 | integer | `uint8_t`  | part of base header |
+| `version`        | 1 | integer | `uint8_t`  | part of base header|
+| `flags`          | 2 | integer | `uint16_t` | part of base header |
+| `rma_iov_count`  | 4 | integer | `uint32_t` | number of RMA iov structure |
+| `rma_iov`        | `rma_iov_count` * 24 | array of `efa_rma_iov` | `efa_rma_iov[]` | remote buffer information |
+
+One thing worth noting is that there is no `msg_id` in the eager RTW header, because EFA provider does not support
+ordered write operation.
+
+#### emulated long-CTS write feature/sub-protocol
+
+emulated long-CTS write sub-protocol is used when the buffer size is too big to fit in one packet.
+
+The workflow of emulated long-CTS write general follow the long-CTS message sub-protocol, as illustrated
+in the following diagram:
+
+![emulated long-CTS write](write_longcts.png)
+
+The main difference between the two protocol is that the LONGCTS_RTW packet is used instead of the
+LONGCTS_RTM packet. The binary format of LONGCTS_RTW packet's mandatory header is listed in table 3.9:
+
+Table: 3.9 the format of LONGCTS_RTW packet's mandatory header
+
+| Name | Length (bytes) | type | C language type | Note |
+|-|-|-|-|-|
+| `type`           | 1 | integer | `uint8_t`  | part of base header |
+| `version`        | 1 | integer | `uint8_t`  | part of base header|
+| `flags`          | 2 | integer | `uint16_t` | part of base header |
+| `rma_iov_count`  | 4 | integer | `uint32_t` | number of RMA iov structure |
+| `msg_length`     | 8 | integer | `uint64_t` | total length of the application buffer |
+| `send_id`        | 4 | integer | `uint32_t` | ID of send operation |
+| `credit_request` | 4 | integer | `uint32_t` | number of packets requester is ready to send |
+| `rma_iov`        | `rma_iov_count` * 24 | array of `efa_rma_iov` | `efa_rma_iov[]` | remote buffer information |
+
+All fields have been described before, but some explanation is warranted for the `send_id` field. It is not
+named `write_id` because this protocol is using send/receive to emulated write, therefore it is implied that
+the requester is treating this communication as a send operation internally, and this communication is subject
+to same flow control as a long-CTS message communication does.
+
+#### emulated read features/sub-protocols
+
+This section describes two emulated read sub-protocols: emulated short read and emulated long-CTS read. Both
+sub-protocols use send/receive to emulate read. The interesting part is, in an emulated read communication,
+the responder is the sender and the requester is the receiver.
+
+The workflow of emulated short read protocol is illustrated in the following diagram:
```

Comment:
I don't believe it's been explicitly noted when this is allowed or used vs long. Although information on selection is deliberately not included, it must be less than the MTU - header size.

Reply:
In the new revision, I added a discussion later in the section.